In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as py

In [14]:
basketdf = pd.read_csv('Data/customer_supermarket.csv', sep='\t', index_col=0, decimal=',')


In [15]:
basketdf.head(10)

BasketID      BasketDate  Sale CustomerID CustomerCountry  ProdID  \
0   536365  01/12/10 08:26  2.55    17850.0  United Kingdom  85123A   
1   536365  01/12/10 08:26  3.39    17850.0  United Kingdom   71053   
2   536365  01/12/10 08:26  2.75    17850.0  United Kingdom  84406B   
3   536365  01/12/10 08:26  3.39    17850.0  United Kingdom  84029G   
4   536365  01/12/10 08:26  3.39    17850.0  United Kingdom  84029E   
5   536365  01/12/10 08:26  7.65    17850.0  United Kingdom   22752   
6   536365  01/12/10 08:26  4.25    17850.0  United Kingdom   21730   
7   536366  01/12/10 08:28  1.85    17850.0  United Kingdom   22633   
8   536366  01/12/10 08:28  1.85    17850.0  United Kingdom   22632   
9   536368  01/12/10 08:34  4.25    13047.0  United Kingdom   22960   

                             ProdDescr  Qta  
0   WHITE HANGING HEART T-LIGHT HOLDER    6  
1                  WHITE METAL LANTERN    6  
2       CREAM CUPID HEARTS COAT HANGER    8  
3  KNITTED UNION FLAG HOT WATER BOTTLE    6  
4       RED WOOLLY HOTTIE WHITE HEART.    6  
5         SET 7 BABUSHKA NESTING BOXES    2  
6    GLASS STAR FROSTED T-LIGHT HOLDER    6  
7               HAND WARMER UNION JACK    6  
8            HAND WARMER RED POLKA DOT    6  
9             JAM MAKING SET WITH JARS    6

In [16]:
basketdf.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 471910 entries, 0 to 541909
Data columns (total 8 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   BasketID         471910 non-null  object 
 1   BasketDate       471910 non-null  object 
 2   Sale             471910 non-null  float64
 3   CustomerID       406830 non-null  object 
 4   CustomerCountry  471910 non-null  object 
 5   ProdID           471910 non-null  object 
 6   ProdDescr        471157 non-null  object 
 7   Qta              471910 non-null  int64  
dtypes: float64(1), int64(1), object(6)
memory usage: 32.4+ MB


In [17]:
basketdf.shape

(471910, 8)

There are object and numeric datas in the dataset. To analyse the customers on amount spend in transactions an amount column is to be added. The basket date is to be converted to date time format, since it is object type

In [18]:
# convert dates in a suitable format
basketdf['BasketDate'] = pd.to_datetime(basketdf['BasketDate'], dayfirst=True)

In [19]:
# add 'Amount' attribute, useful for some additional features on customer's purchase behaviour
basketdf['Amount'] = basketdf['Sale'] * basketdf['Qta']

In [20]:
basketdf.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 471910 entries, 0 to 541909
Data columns (total 9 columns):
 #   Column           Non-Null Count   Dtype         
---  ------           --------------   -----         
 0   BasketID         471910 non-null  object        
 1   BasketDate       471910 non-null  datetime64[ns]
 2   Sale             471910 non-null  float64       
 3   CustomerID       406830 non-null  object        
 4   CustomerCountry  471910 non-null  object        
 5   ProdID           471910 non-null  object        
 6   ProdDescr        471157 non-null  object        
 7   Qta              471910 non-null  int64         
 8   Amount           471910 non-null  float64       
dtypes: datetime64[ns](1), float64(2), int64(1), object(5)
memory usage: 36.0+ MB


In [21]:
basketdf.shape

(471910, 9)

Checking for null values

In [22]:
basketdf.isnull().sum()

BasketID               0
BasketDate             0
Sale                   0
CustomerID         65080
CustomerCountry        0
ProdID                 0
ProdDescr            753
Qta                    0
Amount                 0
dtype: int64

ProdDescr and CustomerId columns have null values. The prodDescr is included in the customerId.
Counting all the rows where both attributes are null

In [23]:
basketdf[basketdf['CustomerID'].isnull() & basketdf['ProdDescr'].isnull()].count()

BasketID           753
BasketDate         753
Sale               753
CustomerID           0
CustomerCountry    753
ProdID             753
ProdDescr            0
Qta                753
Amount             753
dtype: int64

In [25]:
basketdf.describe().T

count       mean         std        min   25%    50%    75%  \
Sale    471910.0   4.030945   83.769380  -11062.06  1.25   2.08   3.75   
Qta     471910.0  10.716533  231.355136  -80995.00  1.00   4.00  12.00   
Amount  471910.0  19.030258  400.925995 -168469.60  3.75  10.08  17.70   

             max  
Sale     38970.0  
Qta      80995.0  
Amount  168469.6

There are negative values in sale, qta and hence amount.
Hence analysing each feature

In [28]:
#BasketID
basketcounts = basketdf['BasketID'].value_counts()
basketcounts.describe()

count    24627.000000
mean        19.162302
std         28.635400
min          1.000000
25%          2.000000
50%         11.000000
75%         24.000000
max        542.000000
Name: BasketID, dtype: float64

The number of distinct baskets is 24627

#Quantity
counting the negative values

In [29]:
# rows with negative qta
qtanegqta = basketdf['Qta'] < 0
basketdf[qtanegqta].count()

BasketID           9752
BasketDate         9752
Sale               9752
CustomerID         8905
CustomerCountry    9752
ProdID             9752
ProdDescr          9305
Qta                9752
Amount             9752
dtype: int64

In [31]:
#cost
# rows with zero cost
zerocost = basketdf['Sale'] == 0
basketdf[zerocost].count()

BasketID           1279
BasketDate         1279
Sale               1279
CustomerID           40
CustomerCountry    1279
ProdID             1279
ProdDescr           526
Qta                1279
Amount             1279
dtype: int64

There are 1279 rows with zero cost